File: services/documents/ot_dedup.py
"""Б7: поиск однотипных инструкций по охране труда.

Отдел ОТ сокращает 448 действующих инструкций: нужно найти пары документов,
у которых «более 80% текста совпадает». Двухступенчатый алгоритм без LLM:

1) дешёвый кандидат-отбор: шинглы (5-словные n-граммы, хеши) + инвертированный
   индекс → Dice-коэффициент по множествам шинглов для всех пар сразу;
2) точная похожесть для кандидатов (Dice ≥ 0.5): difflib.SequenceMatcher
   по СЛОВАМ (быстро и близко к человеческому «процент совпадения текста»).

Пары ≥ порога объединяются в группы (связные компоненты) — так видно целые
семейства однотипных инструкций, а не только парные совпадения.
"""

from __future__ import annotations

import io
import re
import zlib
from collections import defaultdict
from difflib import SequenceMatcher

from utils.logger import logger

# Триггер чат-команды: «найди дубликаты/однотипные инструкции»
OT_DEDUP_REQUEST_RE = re.compile(
    r"(?:дубл|однотип|совпаден)[\w-]*[^.]{0,60}инструкц|инструкц[\w-]*[^.]{0,60}(?:дубл|однотип)",
    re.IGNORECASE,
)

_WORD_RE = re.compile(r"[а-яёa-z0-9]+", re.IGNORECASE)
_SHINGLE = 5
# Кандидаты на точное сравнение (быстрый Dice по шинглам)
_CANDIDATE_DICE = 0.5
# Порог «однотипные» из брифа отдела ОТ
DUPLICATE_THRESHOLD = 0.80
# Пары ниже порога, но выше этого — «пограничные», тоже попадают в отчёт
REPORT_THRESHOLD = 0.60


def _words(text: str) -> list[str]:
    return _WORD_RE.findall((text or "").lower())


def _shingles(words: list[str]) -> set[int]:
    if len(words) < _SHINGLE:
        return {zlib.crc32(" ".join(words).encode("utf-8"))} if words else set()
    return {
        zlib.crc32(" ".join(words[i:i + _SHINGLE]).encode("utf-8"))
        for i in range(len(words) - _SHINGLE + 1)
    }


def compare_documents(docs: list[tuple[str, str]]) -> dict:
    """docs: [(имя файла, текст)] → {pairs, groups, files}.

    pairs: [{a, b, percent}] по убыванию похожести (percent — пословный
    SequenceMatcher, ≥ REPORT_THRESHOLD); groups: связные компоненты пар
    ≥ DUPLICATE_THRESHOLD с диапазоном похожести внутри группы.
    """
    names = [n for n, _ in docs]
    word_lists = [_words(t) for _, t in docs]
    shingle_sets = [_shingles(w) for w in word_lists]

    # Инвертированный индекс шинглов → счётчик общих шинглов по парам
    by_shingle: dict[int, list[int]] = defaultdict(list)
    for i, ss in enumerate(shingle_sets):
        for h in ss:
            by_shingle[h].append(i)
    common: dict[tuple[int, int], int] = defaultdict(int)
    for ids in by_shingle.values():
        if len(ids) < 2 or len(ids) > 50:  # шингл-«шум», общий для всех, не информативен
            continue
        for x in range(len(ids)):
            for y in range(x + 1, len(ids)):
                common[(ids[x], ids[y])] += 1

    candidates: list[tuple[int, int]] = []
    for (i, j), c in common.items():
        denom = len(shingle_sets[i]) + len(shingle_sets[j])
        if denom and (2 * c / denom) >= _CANDIDATE_DICE:
            candidates.append((i, j))
    logger.info("[OT-DEDUP] файлов: {}, пар-кандидатов: {}", len(docs), len(candidates))

    pairs: list[dict] = []
    for i, j in candidates:
        ratio = SequenceMatcher(a=word_lists[i], b=word_lists[j], autojunk=False).ratio()
        if ratio >= REPORT_THRESHOLD:
            pairs.append({
                "a": names[i], "b": names[j],
                "percent": round(ratio * 100, 1),
                "_i": i, "_j": j,
            })
    pairs.sort(key=lambda p: -p["percent"])

    # Группы однотипных (связные компоненты по парам ≥ 80%)
    parent = list(range(len(docs)))

    def find(x: int) -> int:
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    for p in pairs:
        if p["percent"] >= DUPLICATE_THRESHOLD * 100:
            a, b = find(p["_i"]), find(p["_j"])
            if a != b:
                parent[a] = b

    members: dict[int, list[int]] = defaultdict(list)
    for i in range(len(docs)):
        members[find(i)].append(i)
    groups: list[dict] = []
    for ids in members.values():
        if len(ids) < 2:
            continue
        in_group = [
            p["percent"] for p in pairs
            if p["percent"] >= DUPLICATE_THRESHOLD * 100
            and find(p["_i"]) == find(ids[0])
        ]
        groups.append({
            "files": sorted(names[i] for i in ids),
            "size": len(ids),
            "min_percent": min(in_group) if in_group else 0,
            "max_percent": max(in_group) if in_group else 0,
        })
    groups.sort(key=lambda g: -g["size"])

    for p in pairs:
        p.pop("_i"), p.pop("_j")
    return {
        "files": len(docs),
        "pairs": pairs,
        "duplicates": len([p for p in pairs if p["percent"] >= DUPLICATE_THRESHOLD * 100]),
        "groups": groups,
    }


def build_dedup_xlsx(result: dict) -> bytes:
    """xlsx-отчёт: лист «Пары» (все ≥60%) + лист «Группы однотипных» (≥80%)."""
    from openpyxl import Workbook
    from openpyxl.styles import Alignment, Font, PatternFill

    wb = Workbook()
    ws = wb.active
    ws.title = "Пары"
    ws.append(["Инструкция 1", "Инструкция 2", "Совпадение, %"])
    for c in ws[1]:
        c.font = Font(bold=True)
    red = PatternFill("solid", start_color="FDE8E8")
    for p in result["pairs"]:
        ws.append([p["a"], p["b"], p["percent"]])
        if p["percent"] >= DUPLICATE_THRESHOLD * 100:
            for c in ws[ws.max_row]:
                c.fill = red
    ws.column_dimensions["A"].width = 60
    ws.column_dimensions["B"].width = 60
    ws.column_dimensions["C"].width = 16

    ws2 = wb.create_sheet("Группы однотипных")
    ws2.append(["Группа", "Файлов", "Совпадение, %", "Файлы (кандидаты на объединение)"])
    for c in ws2[1]:
        c.font = Font(bold=True)
    for gi, g in enumerate(result["groups"], 1):
        rng = (
            f"{g['min_percent']:.0f}–{g['max_percent']:.0f}"
            if g["min_percent"] != g["max_percent"] else f"{g['max_percent']:.0f}"
        )
        ws2.append([gi, g["size"], rng, "\n".join(g["files"])])
        ws2.cell(ws2.max_row, 4).alignment = Alignment(wrap_text=True, vertical="top")
    ws2.column_dimensions["D"].width = 90

    buf = io.BytesIO()
    wb.save(buf)
    return buf.getvalue()


def run_dedup_zip(db, user, zip_path) -> tuple:
    """Полный цикл для ZIP с инструкциями: распаковка → парсинг → сравнение →
    xlsx-отчёт в «Мои документы». Используется API-эндпоинтом и чат-веткой.
    Возвращает (MyDocuments, result)."""
    import tempfile
    import zipfile
    from datetime import datetime
    from pathlib import Path

    from config import settings
    from data.my_documents import MyDocuments
    from services.parsers import parse_file

    allowed = {".docx", ".doc", ".pdf", ".rtf", ".txt", ".odt"}
    zf = zipfile.ZipFile(zip_path)
    docs: list[tuple[str, str]] = []
    errors: list[str] = []
    with tempfile.TemporaryDirectory(prefix="ot_dedup_") as tmpdir:
        for info in zf.infolist():
            if info.is_dir() or len(docs) >= 500:
                continue
            fname = Path(info.filename).name
            suffix = Path(fname).suffix.lower()
            if not fname or suffix not in allowed:
                continue
            p = Path(tmpdir) / f"{len(docs)}{suffix}"
            p.write_bytes(zf.read(info))
            try:
                parsed = parse_file(p)
                if (parsed.text or "").strip():
                    docs.append((fname, parsed.text))
                else:
                    errors.append(fname)
            except Exception as e:
                logger.warning("[OT-DEDUP] {} не распарсился: {}", fname, e)
                errors.append(fname)
    if len(docs) < 2:
        raise ValueError("В архиве меньше двух читаемых инструкций (docx/doc/pdf/rtf/txt)")

    result = compare_documents(docs)
    result["unreadable"] = errors

    settings.docs_generated.mkdir(parents=True, exist_ok=True)
    out = settings.docs_generated / f"ot_dedup_{datetime.now().strftime('%Y%m%d_%H%M%S')}.xlsx"
    out.write_bytes(build_dedup_xlsx(result))
    rec = MyDocuments(
        user_id=user.id,
        title=f"Дубликаты инструкций ОТ ({result['files']} файлов, {result['duplicates']} пар ≥80%)",
        template_key="ot_dedup",
        file_path=str(out),
        progress=100,
        status="ready",
        fields={"files": result["files"], "duplicates": result["duplicates"]},
    )
    db.add(rec)
    db.commit()
    db.refresh(rec)
    return rec, result

File: app/lib/docs/ot-dedup.ts
import 'server-only';
import { Prisma } from '@prisma/client';
import { prisma } from '@/lib/db';
import { buildXlsx, type XlsxCell } from './xlsx';
import { crc32, sequenceRatio } from './seqmatch';
import { DocValueError, round1, saveGenerated, timestamp } from './common';
import { toDocsPath } from '@/lib/news';

/**
 * Б7: поиск однотипных инструкций по охране труда.
 * Порт services/documents/ot_dedup.py. Без LLM.
 *
 * 1) дешёвый кандидат-отбор: шинглы (5-словные n-граммы, crc32) + инвертированный
 *    индекс → Dice-коэффициент по множествам шинглов для всех пар сразу;
 * 2) точная похожесть для кандидатов: SequenceMatcher по СЛОВАМ.
 */

/**
 * Триггер чат-команды «найди дубликаты инструкций».
 * `[\w-]` из Python пишем классом целиком: ru() подставляет \w как готовый
 * класс, и вложить его внутрь другого класса нельзя.
 */
export const OT_DEDUP_REQUEST_RE =
  /(?:дубл|однотип|совпаден)[0-9A-Za-zА-Яа-яЁё_-]*[^.]{0,60}инструкц|инструкц[0-9A-Za-zА-Яа-яЁё_-]*[^.]{0,60}(?:дубл|однотип)/i;

const WORD_RE = /[а-яёa-z0-9]+/gi;
const SHINGLE = 5;
const CANDIDATE_DICE = 0.5;
/** Порог «однотипные» из брифа отдела ОТ. */
export const DUPLICATE_THRESHOLD = 0.8;
/** Пары ниже порога, но выше этого — «пограничные», тоже попадают в отчёт. */
const REPORT_THRESHOLD = 0.6;

function words(text: string): string[] {
  return (text || '').toLowerCase().match(WORD_RE) ?? [];
}

function shingles(ws: string[]): Set<number> {
  const enc = new TextEncoder();
  if (ws.length < SHINGLE) {
    return ws.length ? new Set([crc32(enc.encode(ws.join(' ')))]) : new Set();
  }
  const out = new Set<number>();
  for (let i = 0; i <= ws.length - SHINGLE; i += 1) {
    out.add(crc32(enc.encode(ws.slice(i, i + SHINGLE).join(' '))));
  }
  return out;
}

export interface DedupPair {
  a: string;
  b: string;
  percent: number;
}
export interface DedupGroup {
  files: string[];
  size: number;
  min_percent: number;
  max_percent: number;
}
export interface DedupResult {
  files: number;
  pairs: DedupPair[];
  duplicates: number;
  groups: DedupGroup[];
  unreadable: string[];
}

/** docs: [имя файла, текст] → пары и группы однотипных. */
export function compareDocuments(docs: [string, string][]): Omit<DedupResult, 'unreadable'> {
  const names = docs.map(([n]) => n);
  const wordLists = docs.map(([, t]) => words(t));
  const shingleSets = wordLists.map(shingles);

  // Инвертированный индекс шинглов → счётчик общих шинглов по парам
  const byShingle = new Map<number, number[]>();
  shingleSets.forEach((ss, i) => {
    for (const h of ss) {
      const list = byShingle.get(h);
      if (list) list.push(i);
      else byShingle.set(h, [i]);
    }
  });
  const common = new Map<string, number>();
  for (const ids of byShingle.values()) {
    // Шингл-«шум», общий для всех файлов, не информативен.
    if (ids.length < 2 || ids.length > 50) continue;
    for (let x = 0; x < ids.length; x += 1) {
      for (let y = x + 1; y < ids.length; y += 1) {
        const key = `${ids[x]},${ids[y]}`;
        common.set(key, (common.get(key) ?? 0) + 1);
      }
    }
  }

  const candidates: [number, number][] = [];
  for (const [key, c] of common) {
    const [i, j] = key.split(',').map(Number);
    const denom = shingleSets[i].size + shingleSets[j].size;
    if (denom && (2 * c) / denom >= CANDIDATE_DICE) candidates.push([i, j]);
  }

  const pairs: (DedupPair & { _i: number; _j: number })[] = [];
  for (const [i, j] of candidates) {
    const ratio = sequenceRatio(wordLists[i], wordLists[j]);
    if (ratio >= REPORT_THRESHOLD) {
      pairs.push({ a: names[i], b: names[j], percent: round1(ratio * 100), _i: i, _j: j });
    }
  }
  pairs.sort((x, y) => y.percent - x.percent);

  // Группы однотипных (связные компоненты по парам ≥ 80%)
  const parent = docs.map((_, i) => i);
  const find = (x: number): number => {
    let cur = x;
    while (parent[cur] !== cur) {
      parent[cur] = parent[parent[cur]];
      cur = parent[cur];
    }
    return cur;
  };
  for (const p of pairs) {
    if (p.percent >= DUPLICATE_THRESHOLD * 100) {
      const a = find(p._i);
      const b = find(p._j);
      if (a !== b) parent[a] = b;
    }
  }

  const members = new Map<number, number[]>();
  for (let i = 0; i < docs.length; i += 1) {
    const root = find(i);
    const list = members.get(root);
    if (list) list.push(i);
    else members.set(root, [i]);
  }
  const groups: DedupGroup[] = [];
  for (const ids of members.values()) {
    if (ids.length < 2) continue;
    const inGroup = pairs
      .filter((p) => p.percent >= DUPLICATE_THRESHOLD * 100 && find(p._i) === find(ids[0]))
      .map((p) => p.percent);
    groups.push({
      files: ids.map((i) => names[i]).sort((a, b) => (a < b ? -1 : a > b ? 1 : 0)),
      size: ids.length,
      min_percent: inGroup.length ? Math.min(...inGroup) : 0,
      max_percent: inGroup.length ? Math.max(...inGroup) : 0,
    });
  }
  groups.sort((a, b) => b.size - a.size);

  return {
    files: docs.length,
    pairs: pairs.map(({ a, b, percent }) => ({ a, b, percent })),
    duplicates: pairs.filter((p) => p.percent >= DUPLICATE_THRESHOLD * 100).length,
    groups,
  };
}

/** «{v:.0f}» из Python: половина округляется к чётному. */
function fmt0(v: number): string {
  const fl = Math.floor(v);
  const diff = v - fl;
  const r = diff > 0.5 ? fl + 1 : diff < 0.5 ? fl : fl % 2 === 0 ? fl : fl + 1;
  return String(r);
}

/** xlsx-отчёт: лист «Пары» (все ≥60%) + лист «Группы однотипных» (≥80%). */
export function buildDedupXlsx(result: Omit<DedupResult, 'unreadable'>): Buffer {
  const pairCells: XlsxCell[] = [];
  ['Инструкция 1', 'Инструкция 2', 'Совпадение, %'].forEach((h, i) => {
    pairCells.push({ row: 1, col: i + 1, value: h, style: { bold: true } });
  });
  result.pairs.forEach((p, idx) => {
    const row = idx + 2;
    const fill = p.percent >= DUPLICATE_THRESHOLD * 100 ? 'FDE8E8' : undefined;
    pairCells.push({ row, col: 1, value: p.a, style: fill ? { fill } : undefined });
    pairCells.push({ row, col: 2, value: p.b, style: fill ? { fill } : undefined });
    pairCells.push({ row, col: 3, value: p.percent, style: fill ? { fill } : undefined });
  });

  const groupCells: XlsxCell[] = [];
  ['Группа', 'Файлов', 'Совпадение, %', 'Файлы (кандидаты на объединение)'].forEach((h, i) => {
    groupCells.push({ row: 1, col: i + 1, value: h, style: { bold: true } });
  });
  result.groups.forEach((g, gi) => {
    const row = gi + 2;
    const rng =
      g.min_percent !== g.max_percent
        ? `${fmt0(g.min_percent)}–${fmt0(g.max_percent)}`
        : fmt0(g.max_percent);
    groupCells.push({ row, col: 1, value: gi + 1 });
    groupCells.push({ row, col: 2, value: g.size });
    groupCells.push({ row, col: 3, value: rng });
    groupCells.push({
      row,
      col: 4,
      value: g.files.join('\n'),
      style: { wrap: true, vAlign: 'top' },
    });
  });

  return buildXlsx([
    {
      name: 'Пары',
      cells: pairCells,
      cols: [{ col: 1, width: 60 }, { col: 2, width: 60 }, { col: 3, width: 16 }],
    },
    { name: 'Группы однотипных', cells: groupCells, cols: [{ col: 4, width: 90 }] },
  ]);
}

/** Полный цикл: разобранные инструкции → сравнение → xlsx в «Мои документы». */
export async function runDedup(userId: number, docs: [string, string][], unreadable: string[]) {
  if (docs.length < 2) {
    throw new DocValueError('В архиве меньше двух читаемых инструкций (docx/doc/pdf/rtf/txt)');
  }
  const result: DedupResult = { ...compareDocuments(docs), unreadable };
  const filePath = await saveGenerated(`ot_dedup_${timestamp()}.xlsx`, buildDedupXlsx(result));
  const rec = await prisma.my_documents.create({
    data: {
      user_id: userId,
      title: `Дубликаты инструкций ОТ (${result.files} файлов, ${result.duplicates} пар ≥80%)`,
      template_key: 'ot_dedup',
      file_path: toDocsPath(filePath),
      progress: 100,
      status: 'ready',
      fields: { files: result.files, duplicates: result.duplicates } as Prisma.InputJsonValue,
      is_pii: false, // значение по умолчанию модели MyDocuments
    },
  });
  return { rec, result };
}
